app/controllers/physiqual/oauth_session_controller.rb

```diff
@@ -65,35 +65,49 @@ def current_user
     end
 
     def check_token
-      my_tokens = current_user.physiqual_tokens
-      if my_tokens.blank? || my_tokens.first.token.blank?
+      # my_tokens = current_user.physiqual_tokens
+      my_tokens = provider_tokens(params[:state])
+
+      if my_tokens.blank? || !my_tokens.first.complete?
         redirect_to authorize_oauth_session_index_path(provider: params[:state])
       else
-        my_tokens.each { |token| token.refresh! if token.expired? && token.complete? }
-        @token = my_tokens.first
+        # Refresh all tokens here?
+        current_user.physiqual_tokens.each { |token| token.refresh! if token.expired? && token.complete? }
       end
     end
 
     def set_token
-      if params[:provider] == GoogleToken.csrf_token
-        @token = current_user.google_tokens.create
-      elsif params[:provider] == FitbitToken.csrf_token
-        @token = current_user.fitbit_tokens.create
-      else
-        head 404
-      end
+      tokens = provider_tokens params[:provider]
+      head 404 and return if tokens.nil?
+      @token = get_or_create_token(tokens)
     end
 
     def token
       @token = current_user.physiqual_tokens.select { |x| x.class.csrf_token == params[:provider] }
       head 404 if @token.blank?
+
+      # @token should always have length == 1
       @token = @token.first
     end
 
-    private
+    def provider_tokens(provider)
+      tokens = nil
+      if provider == GoogleToken.csrf_token
+        tokens = current_user.google_tokens
+      elsif provider == FitbitToken.csrf_token
+        tokens = current_user.fitbit_tokens
+      end
+      tokens
+    end
+
+    def get_or_create_token(tokens)
+      return tokens.first unless tokens.blank?
+      tokens.create
```

Comment:
Dit geeft volgens mij een error als tokens nil is dwz als de params[:provider] niet google of fitbit is. Kunnen we daar een 404 van maken?